src/react/hooks/use-messages.ts

```diff
@@ -0,0 +1,130 @@
+   * up until the listener was attached, in the oldest to newest order.
+   *
+   * It is advised to call this method after any discontinuity event; to retrieve messages that may have been missed
+   * before the listener was re-attached.
+   *
+   * @param options - The query options to use when fetching the previous messages.
+   *
+   * @defaultValue - This will be undefined if no listener is provided in the {@link UseMessagesParams}.
+   */
+  readonly getPreviousMessages?: MessageSubscriptionResponse['getPreviousMessages'];
+}
+
+export interface UseMessagesParams extends StatusParams, Listenable<MessageListener> {
+  /**
+   * An optional listener that can be provided to receive new messages in the room.
+   */
+  listener?: MessageListener;
+}
+
+/**
+ * A hook that provides access to the {@link Messages} instance in the room.
+ * It will use the instance belonging to the room in the nearest {@link ChatRoomProvider} in the component tree.
+ * If a listener is provided, it will subscribe to new messages in the room,
+ * and will also set the {@link getPreviousMessages}.
+ *
+ * @param params - Allows the registering of optional callbacks.
+ * @returns UsePresenceResponse - An object containing the {@link Messages} instance and methods to interact with it.
+ */
+export const useMessages = (params?: UseMessagesParams): UseMessagesResponse => {
+  const { currentStatus: connectionStatus, error: connectionError } = useChatConnection({
+    onStatusChange: params?.onConnectionStatusChange,
+  });
+  const { room, roomError, roomStatus } = useRoom({
+    onStatusChange: params?.onRoomStatusChange,
+  });
+
+  const logger = useLogger();
+  logger.trace('useMessages();', { params, roomId: room.roomId });
+
+  // we are storing the params in a ref so that we don't end up with an infinite loop should the user pass
+  // in an unstable reference
+  const paramsRef = useRef(params);
+
+  useLayoutEffect(() => {
+    paramsRef.current = params;
+  });
+
+  const send = useCallback((params: SendMessageParams) => room.messages.send(params), [room]);
+  const get = useCallback((options: QueryOptions) => room.messages.get(options), [room]);
+
+  const [getPreviousMessages, setGetPreviousMessages] = useState<MessageSubscriptionResponse['getPreviousMessages']>();
+
+  useEffect(() => {
+    if (paramsRef.current?.listener) {
+      logger.debug('useMessages(); applying listener', { roomId: room.roomId });
+      const sub = room.messages.subscribe(paramsRef.current.listener);
+
+      // set the getPreviousMessages method if a listener is provided
+      setGetPreviousMessages(() => {
+        logger.debug('useMessages(); setting getPreviousMessages state', { roomId: room.roomId });
+        return (params: Omit<QueryOptions, 'direction'>) => {
+          return sub.getPreviousMessages(params);
+        };
+      });
+
+      return () => {
+        logger.debug('useMessages(); removing listener', { roomId: room.roomId });
```

Comment:
nitpick: should this and the log on line 103 be a single log? or trace one of them or both? seems like a lot of logging

Reply:
Good point, I'll combined this to a single log